### PR TITLE
Add form submission models

### PR DESCRIPTION
**Pull request by Karl456**

### CODE REVIEW

File: api-reference/openapi.yaml [hunk 1]
```diff
@@ -2735,3 +2735,159 @@ components:
...
+            content: 
+              type: object
+              description: The content of the submission
+        meta:
+          type: object
+          properties:
+            created_at:
+              type: string
+              example: 2024-06-13T07:15:23Z
+              description: "**Sortable**. The time the operation was created ([RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339))"
```
Comment (andrew-waters) [inline api-reference/openapi.yaml]: not sure this is the right description

File: api-reference/openapi.yaml [hunk 2]
```diff
@@ -2735,3 +2735,159 @@ components:
             self:
               type: string
               example: https://api.ctrl-hub.com/v3/governance/schemes/5852cf16-8ec8-4aba-809e-f8d828772c4e/work-orders/5852cf16-8ec8-4aba-809e-f8d828772c4e/permits/5852cf16-8ec8-4aba-809e-f8d828772c4e
+    dataCaptureFormSubmissionVersionModel:
+      type: object
+      properties:
+        type:
+          type: string
+          example: versions
```
Comment (andrew-waters) [inline api-reference/openapi.yaml]: Should this be `submission-versions` to disambiguate with things like form versions?

File: api-reference/openapi.yaml [hunk 3]
```diff
@@ -2735,3 +2735,159 @@ components:
...
+        attributes:
+          type: object
+          properties:
+            status:
+              type: string
+              description: The status of the form submission
+              example: complete
+            content: 
+              type: object
+              description: The content of the submission
```
Comment (andrew-waters) [inline api-reference/openapi.yaml]: let's adda little bit of example content in here, such as `field: value`

File: api-reference/openapi.yaml [hunk 4]
```diff
@@ -2735,3 +2735,159 @@ components:
...
+          type: object
+          properties:
+            created_at:
+              type: string
+              example: 2024-06-13T07:15:23Z
+              description: "**Sortable**. The time the operation was created ([RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339))"
+        relationships:
+          type: object
+          properties:
+            form_version:
```
Comment (andrew-waters) [inline api-reference/openapi.yaml:2916]: we're missing the `form` relationship here

File: api-reference/openapi.yaml [hunk 5]
```diff
@@ -2735,3 +2735,159 @@ components:
...
+            content:
+              type: object
+              description: The content of the form submission
+        meta:
+          type: object
+          properties:
+            created_at:
+              type: string
+              example: 2024-06-13T07:15:23Z
+              description: "**Sortable**. The time the operation was created ([RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339))"
```
Comment (andrew-waters) [inline api-reference/openapi.yaml]: copy pasta

File: api-reference/openapi.yaml [hunk 6]
```diff
@@ -2735,3 +2735,159 @@ components:
...
+            updated_at:
+              type: string
+              example: 2024-06-13T07:15:23Z
+              description: "**Sortable**. The time the operation was updated ([RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339))"
```
Comment (andrew-waters) [inline api-reference/openapi.yaml]: copy pasta

File: api-reference/openapi.yaml [hunk 7]
```diff
@@ -2735,3 +2735,159 @@ components:
...
+              type: string
+              description: The reference of the form submission
+              example: 1234A567BCD
+            status:
+              type: string
+              description: The status of the form submission
+              example: complete
+            content:
+              type: object
+              description: The content of the form submission
```
Comment (andrew-waters) [inline api-reference/openapi.yaml]: Will this pull the content from the latest submission version? Or is it redundant here and instead we rely on API client navigating to `/v3/orgs/{org}/data-capture/submissions/{id}/relationships/versions/latest`?

My preference is for the latter - we can always add an optional `?include=latest` or something similar to pull back the content, but doing so by default will increase the payload on the wire here.

File: api-reference/openapi.yaml [hunk 8]
```diff
@@ -2735,3 +2735,159 @@ components:
...
+          type: object
+          properties:
+            reference:
+              type: string
+              description: The reference of the form submission
+              example: 1234A567BCD
+            status:
+              type: string
+              description: The status of the form submission
+              example: complete
```
Comment (andrew-waters) [inline api-reference/openapi.yaml]: Might be worth making a comment here in the description that the status is always the _latest_ status based on submission versions

File: api-reference/openapi.yaml [hunk 9]
```diff
@@ -2735,3 +2735,159 @@ components:
...
+                  items:
+                    $ref: "#/components/schemas/dataCaptureFormVersionRelationship"
+            author:
+              type: object
+              description: The id of user this submission relates to
+              properties:
+                data:
+                  type: array
+                  items:
+                    $ref: "#/components/schemas/userRelationships"
```
Comment (andrew-waters) [inline api-reference/openapi.yaml]: I'd be tempted to add in here relationships to versions as well - which has a similar effect as call the `.../relationships/versions` endpoint. That way in one call, the API client can see the count of versions the submission has had

File: api-reference/openapi.yaml (hunk 4 above)
Comment (Karl456) [reply api-reference/openapi.yaml:2916]: we only store the form version against a submission version, so we'd have to reach out to mongo to get the form via the version id - do we want to do this? especially on the GetMany pages?

Comment (andrew-waters) [reply api-reference/openapi.yaml:2916]: we should already have the submission document in memory - it's how we get the form version and the content. We should get this for free... unless you're using $project aggregations

Comment (andrew-waters) [reply api-reference/openapi.yaml:2916]: in fact, we should also include a relationship to the submission on the version

**Review by andrew-waters — COMMENTED** [review]
Let's rename FormSubmissions (and other versions of it) to `Submissions`. There's a few instances of it, and I think it makes more sense without the forms part in the way we communicate. Happy to do that if you want me to pick this up

**Comment by Karl456** [comment]
> Let's rename FormSubmissions (and other versions of it) to `Submissions`. There's a few instances of it, and I think it makes more sense without the forms part in the way we communicate. Happy to do that if you want me to pick this up

Just pushed - let me know if I missed any references

**Comment by andrew-waters** [comment]
👍🏼  just the ones in the api spec

**Review by andrew-waters — APPROVED** [review]

**Pull request merged**